### ISSUE TITLE
Rule proposal: Prefer lookahead and lookbehind

### ISSUE THREAD

**Issue by fisker** [ota-meshi/eslint-plugin-regexp#183]
Avoid capture unneeded groups.

## Fail

```js
'I love unicorn! I hate unicorn?'.replace(/(?<before>love )unicorn(?<after>!)/, '$<before>🦄$<after>');
```

```js
'I love unicorn! I hate unicorn?'.replace(/(love )unicorn(!)/, '$1🦄$2');
```

## Pass

```js
'I love unicorn! I hate unicorn?'.replace(/(?<=love )unicorn(?=!)/, '🦄');
```

Original posted in https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1113, can we implement it here?

**Comment by RunDevelopment** [comment]
Interesting rule, but we have to be careful when implementing this. 

The problem is that capturing groups consume parts of the string while lookarounds do not. This can lead to regexes behaving differently. Examples:

```js
"aaaaaa".replace(/(a)a(a)/g, "$1b$2")
// 'abaaba'
"aaaaaa".replace(/(?<=a)a(?=a)/g, "b")
// 'abbbba'

"aaaaaa".replace(/(^a+)a(a)/, "$1b$2")
// 'aaaaba'
"aaaaaa".replace(/(?<=^a+)a(?=a)/, "b")
// 'abaaaa'
```

That being said, the rule is still a good idea. We just have to find a way to do it safely.

---

I also want to say that the lookbehind replacement might prevent certain optimizations. IIRC, then v8's Boyer–Moore can't be used when a regex contains lookbehinds like this. I think, I read a v8 issue related to this some time ago but don't quote me on that.

We should definitely do some benchmarks to see whether this transformation negatively impacts performance.
